pontoon/insights/utils.py

```diff
@@ -0,0 +1,120 @@
+from django.db.models.functions import TruncMonth
+from django.db.models import Avg, Sum
+
+from pontoon.base.utils import aware_datetime, convert_to_unix_time, get_last_months
+from pontoon.insights.models import LocaleInsightsSnapshot
+
+
+def get_insights(query_filters=None):
+    """Get data required by the Insights tab.
+
+    :param django.db.models.Q query_filters: filters insights by given query_filters.
+    """
+    months = sorted(
+        [aware_datetime(year, month, 1) for year, month in get_last_months(12)]
+    )
+
+    snapshots = LocaleInsightsSnapshot.objects.filter(created_at__gte=months[0])
+
+    if query_filters:
+        snapshots = snapshots.filter(query_filters)
+
+    insights = (
+        snapshots
+        # Truncate to month and add to select list
+        .annotate(month=TruncMonth("created_at"))
+        # Group By month
+        .values("month")
+        # Select the avg/sum of the grouping
+        .annotate(
+            unreviewed_suggestions_lifespan_avg=Avg("unreviewed_suggestions_lifespan")
+        )
+        .annotate(completion_avg=Avg("completion"))
+        .annotate(human_translations_sum=Sum("human_translations"))
+        .annotate(machinery_translations_sum=Sum("machinery_translations"))
+        .annotate(new_source_strings_sum=Sum("new_source_strings"))
+        .annotate(unreviewed_avg=Avg("unreviewed_strings"))
+        .annotate(peer_approved_sum=Sum("peer_approved"))
+        .annotate(self_approved_sum=Sum("self_approved"))
+        .annotate(rejected_sum=Sum("rejected"))
+        .annotate(new_suggestions_sum=Sum("new_suggestions"))
+        # Select month and values
+        .values(
+            "month",
+            "unreviewed_suggestions_lifespan_avg",
+            "completion_avg",
+            "human_translations_sum",
+            "machinery_translations_sum",
+            "new_source_strings_sum",
+            "unreviewed_avg",
+            "peer_approved_sum",
+            "self_approved_sum",
+            "rejected_sum",
+            "new_suggestions_sum",
+        )
+        .order_by("month")
+    )
+
+    dates = [convert_to_unix_time(month) for month in months]
+
+    # Active users
+    total_managers = 0
+    total_reviewers = 0
+    total_contributors = 0
+    active_managers = 0
+    active_reviewers = 0
+    active_contributors = 0
+
+    if snapshots:
+        latest_snapshot = snapshots.latest("created_at")
```

Comment:
How often Pontoon is going to create insight snapshots? I ponder over the reasons behind the denormalization.

Reply:
Once a day.

It's almost impossible to get the exact numbers for the period of the past 12 months from a certain point in time. We simply don't collect all the required information today to tell you what was the overall completion of German for example.

We'll still try to do that as part of https://bugzilla.mozilla.org/show_bug.cgi?id=1680055, but those numbers will be as good estimates as we'll be able to make. They will also be pushed off the charts after 365 days.